supabase/migrations/20230817110845_access_token.sql

```diff
@@ -0,0 +1,174 @@
+        raise exception 'Token with name `%s` already exists', token_name;
+    end;
+
+    return replace(user_id::text, '-', '') || token_text;
+end;
+$$;
+
+create type app.access_token_struct as (
+    id uuid,
+    token_name text,
+    created_at timestamptz
+);
+
+create or replace function public.get_access_tokens()
+    returns setof app.access_token_struct
+    language plpgsql
+    strict
+as $$
+declare
+    user_id uuid = auth.uid();
+    account app.accounts = account from app.accounts account where id = user_id;
+    token bytea = gen_random_bytes(16);
+    token_hash bytea = pgsodium.crypto_pwhash_str(token);
+    token_text text = encode(token, 'hex');
+begin
+    return query
+    select id, token_name, created_at
+    from app.access_tokens at
+    where at.user_id = account.id;
+end;
+$$;
+
+create or replace function public.delete_access_token(
+    token_id uuid
+)
+    returns void
+    language plpgsql
+    strict
+as $$
+declare
+    user_id uuid = auth.uid();
+    account app.accounts = account from app.accounts account where id = user_id;
+begin
+    delete from app.access_tokens at
+    where at.user_id = account.id and at.id = token_id;
+end;
+$$;
+
+create or replace function public.redeem_access_token(
+    access_token text
+)
+    returns text
+    language plpgsql
+    security definer
+    strict
+as $$
+declare
+    user_id_text text;
+    user_id uuid;
+    token_text text;
+    token bytea;
+    access_token_record record;
+    token_hash bytea;
+    token_valid boolean;
+    now timestamp;
+    one_hour_from_now timestamp;
+    issued_at int;
+    expiry_at int;
+    jwt_secret text;
+begin
+    -- validate access token
+    if length(access_token) != 64 then
+        raise exception 'Invalid token';
+    end if;
+
+    user_id_text := substring(access_token from 1 for 32);
+    user_id := substring(user_id_text from 1 for 8) || '-' ||
+                   substring(user_id_text from 9 for 4) || '-' ||
+                   substring(user_id_text from 13 for 4) || '-' ||
+                   substring(user_id_text from 17 for 4) || '-' ||
+                   substring(user_id_text from 21 for 12);
+    token_text := substring(access_token from 33 for 32);
+    token := decode(token_text, 'hex');
+
+    for token_hash in
+        select t.token_hash from app.access_tokens t where t.user_id = user_id
+    loop
+        token_valid := pgsodium.crypto_pwhash_str_verify(token_hash, token);
+        exit when token_valid;
+    end loop;
+
+    if not token_valid then
+        raise exception 'Invalid token';
+    end if;
+
+    -- Generate JWT token
+    now := current_timestamp;
+    one_hour_from_now := now + interval '1 hour';
+    issued_at := date_part('epoch', now);
+    expiry_at := date_part('epoch', one_hour_from_now);
+    jwt_secret := coalesce(
+        current_setting('app.settings.jwt_secret', true),
+        'super-secret-jwt-token-with-at-least-32-characters-long'
```

Comment:
AFAIK in development the `current_setting('app.settings.jwt_secret', true)` [doesn't work during development](https://github.com/supabase/supabase-js/issues/25#issuecomment-1019935888). But I agree that this can be dangerous to fall back to a hardcoded value, so I have removed it.

Ideally, this setting should be returned in development as well. If not, then if I could read another setting which can tell me the current environment will also work.

Reply:
I will release a fix to CLI to set these jwt configs for local development so it matches our hosted platform. https://github.com/supabase/cli/pull/1432